### PR TITLE
realtek-poe: build default config from board.json

### PR DESCRIPTION
**Pull request by Hurricos**
Maintainer: me
Compile tested: (not yet)
Run tested: (not yet)

This adjustment makes a compromise between those that would prefer not to complicate the realtek-poe.git source repository and those who want to have PoE work out-of-the-box on Realtek switches -- see realtek-poe.git, commit 2df137ee457155 ("realtek-poe/openwrt: Generate config from board.json")

Switches previously booted with OpenWrt will have an /etc/config/poe file already present, and thus the uci-default shall not take effect.

### CODE REVIEW

**Comment by Hurricos** [comment]
@howels @mrnuke @blogic

**Review by howels — APPROVED** [review]

File: utils/realtek-poe/Makefile [hunk 1]
```diff
@@ -27,9 +27,10 @@ define Package/realtek-poe/install
 	$(INSTALL_DIR) $(1)/usr/bin
 	$(INSTALL_DIR) $(1)/etc/config
 	$(INSTALL_DIR) $(1)/etc/init.d
+	$(INSTALL_DIR) $(1)/etc/uci-defaults
 	$(INSTALL_BIN) $(PKG_BUILD_DIR)/realtek-poe $(1)/usr/bin/
 	$(INSTALL_BIN) $(PKG_BUILD_DIR)/files/etc/init.d/poe $(1)/etc/init.d/
-	$(INSTALL_CONF) $(PKG_BUILD_DIR)/files/etc/config/poe $(1)/etc/config/
+	$(CP) ./files/etc/uci-defaults/30-poe $(1)/etc/uci-defaults/30-poe
```
Comment (mrnuke) [inline utils/realtek-poe/Makefile:33]: Shouldn't this use some sort of `INSTALL_` macro, so that the file is installed with the correct permissions?

Comment (Hurricos) [reply utils/realtek-poe/Makefile:33]: I pulled the behavior from `libreswan`, but yeah, there are certainly newer changesets showing the use of the `INSTALL_BIN` macro -- e.g. commit 3219c50828914d82da375e959b06da3f6f40d2a4.

Adjusting.

Comment (Hurricos) [reply utils/realtek-poe/Makefile]: Fixed with the push of bb0fb2b7f864a5c0a9e0cf9ffcb05d82e200c21b.

File: utils/realtek-poe/Makefile [hunk 2]
```diff
@@ -27,9 +27,10 @@ define Package/realtek-poe/install
 	$(INSTALL_DIR) $(1)/usr/bin
 	$(INSTALL_DIR) $(1)/etc/config
 	$(INSTALL_DIR) $(1)/etc/init.d
+	$(INSTALL_DIR) $(1)/etc/uci-defaults
 	$(INSTALL_BIN) $(PKG_BUILD_DIR)/realtek-poe $(1)/usr/bin/
 	$(INSTALL_BIN) $(PKG_BUILD_DIR)/files/etc/init.d/poe $(1)/etc/init.d/
-	$(INSTALL_CONF) $(PKG_BUILD_DIR)/files/etc/config/poe $(1)/etc/config/
+	$(INSTALL_BIN) ./files/etc/uci-defaults/30-poe $(1)/etc/uci-defaults/30-poe
 endef
```
Comment (svanheule) [inline utils/realtek-poe/Makefile:34]: The poe-common package also added a section to mark the config file as such. I don't quite remember what this means in terms of behaviour on upgrades etc, and I can't really find much documentation either.

```
define Package/poe-common/conffiles
/etc/config/poe
endef
```

From the opkg code, I think this results in detecting that the file exists and has been "modified" (created), thus doing The Right Thing (creating backup etc.) on package upgrades. I think the old package would just blindly replace the edited poe config file on upgrades. The uci-defaults script would do the same though, and won't touch the file if it already exists.

Hopefully this also means the config file is put in the config backup automatically on system upgrades, but you may have to check this.

**Comment by BKPepe** [comment]
Don't forget to bump PKG_RELEASE.

File: utils/realtek-poe/Makefile (hunk 2 above)
Comment (luizluca) [reply utils/realtek-poe/Makefile:34]: Yes, it matters for the backup. Opkg keeps track of those for hashes to detect if they changed and save them in backups. See list_conffiles() in /sbin/sysupgrade and its usages.

(I wonder how apk change would affect it...)

**Comment by howels** [comment]
@Hurricos are you still working on this PR for the realtek-poe package? I am considering opening another PR myself to close this out.

**Comment by Hurricos** [comment]
Hi! I am still alive.

My underatanding is that only the review feedback blocks the merge.

I think the PKG_RELEASE details eluded me at the time. @howels can you hand me a diff which increments it correctly?

Let's get this merged.

File: utils/realtek-poe/Makefile (hunk 2 above)
Comment (Hurricos) [reply utils/realtek-poe/Makefile:34]: Oh, my bad @howels. This was the other remaining piece of CR feedback. I don't have the setup to test atm, could you build/test how this option affects sysupgrade::list_conffiles() checking for default backup?

**Comment by howels** [comment]
> Hi! I am still alive.
> 
> My underatanding is that only the review feedback blocks the merge.
> 
> I think the PKG_RELEASE details eluded me at the time. @howels can you hand me a diff which increments it correctly?
> 
> Let's get this merged.

This is the field for package release, just increment it to show that the packaging code has been updated:

```
--- a/utils/realtek-poe/Makefile
+++ b/utils/realtek-poe/Makefile
@@ -3,7 +3,7 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=realtek-poe
-PKG_RELEASE:=1
+PKG_RELEASE:=2
 PKG_LICENSE:=GPL-2.0
```

File: utils/realtek-poe/Makefile (hunk 2 above)
Comment (howels) [reply utils/realtek-poe/Makefile:34]: @Hurricos the package manager already seems to preserve the file `/etc/config/poe` when using the master branch with `apk` vs `opkg` without adding this `define` but it does appear to be [common practice in the majority of packages](https://github.com/search?q=repo%3Aopenwrt%2Fopenwrt+%22%2Fconffiles%22&type=code&p=2), so this appears to be best-practice and might be more important for existing `opkg`-based installs. 

I can confirm that `/etc/config/poe` is also included in my backups.

**Comment by howels** [comment]
> Hi! I am still alive.
> 
> My underatanding is that only the review feedback blocks the merge.
> 
> I think the PKG_RELEASE details eluded me at the time. @howels can you hand me a diff which increments it correctly?
> 
> Let's get this merged.

Can you check the diff I posted above to increment the package version? The rest seems to perform correctly based on my tests. 

I created my own versions by incrementing the package ver locally and was able to install these "newer" versions without wiping out my customised `/etc/config/poe` whether that was manually edited or created by your `/etc/uci-defaults/30-poe` script.

**Comment by Hurricos** [comment]
I will have this merged, or actual feedback for what, if anything, needs fixed, by EOD today. Sorry for all the delay.

**Comment by howels** [comment]
> I will have this merged, or actual feedback for what, if anything, needs fixed, by EOD today. Sorry for all the delay.

Thanks 🙂

**Comment by Hurricos** [comment]
9b40724ed increments `PKG_RELEASE` to `2`.

Install-testing on a JG928A on Snapshot now, then gonna try to build the ipk manually to make totally sure it works on our last opkg-enabled release.

File: utils/realtek-poe/Makefile (hunk 2 above)
Comment (BKPepe) [reply utils/realtek-poe/Makefile:34]: @Hurricos have you checked this?

Comment (Hurricos) [reply utils/realtek-poe/Makefile:34]: I'm not sure what's left to check: @howels reports that, despite not being needed, the diff potentially needed to specify conffiles is not actually required for the preservation behavior we want. @BKpepe, do you have suggestions? If no suggestions and just a testing request -- what behavior are we looking for?

It's a bit fuzzy to me.

**Comment by Hurricos** [comment]
I've poked @robimarko on IRC (and now here) to merge this, seeing as he was last to merge a realtek-poe package change from @howels.

File: utils/realtek-poe/Makefile (hunk 2 above)
Comment (Hurricos) [reply utils/realtek-poe/Makefile:34]: Given that @howels testing shows it working under opkg, and mine under apk, I am comfortable not unraveling the mystery right now.

**Comment by Neustradamus** [comment]
Thanks to @Hurricos for this PR and @robimarko for this merging and all others too!

File: utils/realtek-poe/Makefile (hunk 2 above)
Comment (BKPepe) [reply utils/realtek-poe/Makefile:34]: Guys, how did you test it? This is wrong and it is not resolved.

So... let's back what conffiles does.
If you mark configuration file as conffile, thus it is not going to be overwritten by each update and everything was said here.

You can verify it, for example, like this (in this using opkg...):

1. opkg install realtek-poe
2. Manually edit /etc/config/poe, save it
3. opkg install realtek-poe --force-reinstall 
(I think that also opkg install realtek-poe --force-overwrite should work)
Now... your conffile was overwriten by the package and there is default configuration file from the package. 😢 

Basically it was said by @svanheule , @luizluca

Comment (howels) [reply utils/realtek-poe/Makefile:34]: This was tested with apk packages, which did preserve the conf files

Comment (BKPepe) [reply utils/realtek-poe/Makefile:34]: I dont have anything with apk yet and I dont have it with foresable future (maybe something will change, dunno). but I bet that this does not work with opkg, there needs to be defined conffiles.

And also, if apk preserves it, I would like to know how, if it is not set in Makefile.

Comment (howels) [reply utils/realtek-poe/Makefile:34]: Please mention extra fields here to keep it in one place https://github.com/openwrt/packages/pull/26789#issuecomment-2982039538
FYI this is used by the realtek target, which has seen rapid growth since 24.10, therefore many of us are already on snapshot/main. I don't mind adopting correct practices but the documentation is scarce and I wasn't able to observe a difference when testing the main branch.
If this is important due to legacy opkg behaviour then we can fix this omission but the PR was to main and not 24.10.
Please continue the discussion on https://github.com/openwrt/packages/pull/26789
